crates/viewer/re_blueprint_tree/src/blueprint_tree.rs

```diff
@@ -576,11 +642,111 @@ impl BlueprintTree {
             ItemContext::BlueprintTree {
                 filter_session_id: self.filter_state.session_id(),
             },
-            &response,
+            response,
             SelectionUpdateBehavior::UseSelection,
         );
-        self.scroll_to_me_if_needed(ui, &item, &response);
-        ctx.handle_select_hover_drag_interactions(&response, item, true);
+        self.scroll_to_me_if_needed(ui, &item, response);
+        ctx.handle_select_hover_drag_interactions(response, item.clone(), true);
+
+        self.handle_range_selection(ctx, blueprint_tree_data, item, response);
+    }
+
+    /// Handle setting/extending the selection based on shift-clicking.
+    fn handle_range_selection(
+        &mut self,
+        ctx: &ViewerContext<'_>,
+        blueprint_tree_data: &BlueprintTreeData,
+        item: Item,
+        response: &Response,
+    ) {
+        // Early out if we're not being clicked.
+        if !response.clicked() {
+            return;
+        }
+
+        let modifiers = ctx.egui_ctx.input(|i| i.modifiers);
+
+        if modifiers.shift {
+            if self.last_clicked_item.is_some() {
+                let items_iterator = self
+                    .items_in_range(ctx, blueprint_tree_data, &item)
+                    .into_iter()
+                    .map(|item| {
+                        (
+                            item,
+                            Some(ItemContext::BlueprintTree {
+                                filter_session_id: self.filter_state.session_id(),
+                            }),
+                        )
+                    });
+
+                if items_iterator.len() > 0 {
+                    if modifiers.command {
+                        ctx.selection_state.extend_selection(items_iterator);
+                    } else {
+                        ctx.selection_state.set_selection(items_iterator);
+                    }
+                }
+            }
+        } else {
+            self.last_clicked_item = Some(item);
+        }
+    }
+
+    /// Selects a range of items in the blueprint tree.
+    ///
+    /// This method selects all [`Item`]s displayed between the provided shift-clicked item and the
+    /// existing last-clicked item (if any). It takes into account the collapsed state, so only
+    /// actually visible items may be selected.
+    fn items_in_range(
+        &mut self,
+        ctx: &ViewerContext<'_>,
+        blueprint_tree_data: &BlueprintTreeData,
+        shift_clicked_item: &Item,
+    ) -> Vec<Item> {
+        let mut items_in_range = vec![];
+        let mut found_last_clicked_items = false;
+        let mut found_shift_clicked_items = false;
+
+        blueprint_tree_data.visit(|blueprint_tree_item| {
+            let item = blueprint_tree_item.item();
+
+            if Some(&item) == self.last_clicked_item.as_ref() {
```

Comment:
wouldn't it makre more sense to early out on method start if `self.last_clicked_item` is none?